### PR TITLE
Add repositories for entities

### PR DESCRIPTION
**Pull request by VadimZharikov**
rework modules and path logic according to tag model

reg repos

add tags repo

add module review repo

reg services

update module entity to path entity relation

update path repo

add path repo

add module repo

add section repo

add chapter repo

remove generic repo get with predicate method

add generic repo

### CODE REVIEW

File: Client/EnlightenmentApp.DAL/Interfaces/IModuleRepository.cs [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+{
+    public interface IModuleRepository
+    {
+        public Task<IEnumerable<ModuleEntity>> GetModules(CancellationToken ct);
+        public Task<ModuleEntity> GetModuleById(int id, CancellationToken ct);
+        public Task<ModuleEntity> AddModule(ModuleEntity moduleEntity, CancellationToken ct);
+        public Task<ModuleEntity> UpdateModule(ModuleEntity moduleEntity, CancellationToken ct);
+        public Task<bool> DeleteModule(int id, CancellationToken ct);
+        public Task<bool> ModuleExists(ModuleEntity moduleEntity, CancellationToken ct);
+    }
```
Comment (Genresal) [inline Client/EnlightenmentApp.DAL/Interfaces/IModuleRepository.cs:13]: Could it and other interfaces be replaced by some generic interface? The functionality looks the same, so you can specify only special methods for repos interfaces and make them dependent on IGenericRepository

File: Client/EnlightenmentApp.DAL/Repositories/ChapterRepository.cs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+        public async Task<bool> DeleteChapter(int id, CancellationToken ct)
+        {
+            var chapter = await _context.Chapters.FindAsync(id, ct);
+            if (chapter != null)
+            {
+                _context.Chapters.Remove(chapter);
+                await _context.SaveChangesAsync(ct);
+                return true;
+            }
```
Comment (Genresal) [inline Client/EnlightenmentApp.DAL/Repositories/ChapterRepository.cs:35]: not a big deal, but, please, add blank lines after the if statement according to the code convention

File: Client/EnlightenmentApp.DAL/Repositories/ChapterRepository.cs [hunk 3]
```diff
@@ -0,0 +1,72 @@
+﻿using EnlightenmentApp.DAL.DataContext;
+using EnlightenmentApp.DAL.Entities;
+using EnlightenmentApp.DAL.Interfaces;
+using Microsoft.EntityFrameworkCore;
+
+namespace EnlightenmentApp.DAL.Repositories
+{
+    public class ChapterRepository : IChapterRepository
```
Comment (Glipchik) [inline Client/EnlightenmentApp.DAL/Repositories/ChapterRepository.cs:8]: What is the sense of generic repo if you don´t use it? Inherit repo classes from GenericRepository class, please

File: Client/EnlightenmentApp.DAL/Repositories/GenericRepository.cs [hunk 4]
```diff
@@ -0,0 +1,45 @@
+﻿using EnlightenmentApp.DAL.Interfaces;
+using Microsoft.EntityFrameworkCore;
+
+namespace EnlightenmentApp.DAL.Repositories
+{
+    class GenericRepository<TEntity> : IGenericRepository<TEntity> where TEntity : class
+    {
+        private DbContext _context;
```
Comment (Glipchik) [inline Client/EnlightenmentApp.DAL/Repositories/GenericRepository.cs:8]: Make _context and _dbset protected, not private, and get 'em in repos to pass to base constructor

**Review by Glipchik — CHANGES_REQUESTED** [review]

**Review by Genresal — APPROVED** [review]

File: Client/EnlightenmentApp.DAL/Repositories/ChapterRepository.cs [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+using EnlightenmentApp.DAL.Entities;
+using EnlightenmentApp.DAL.Interfaces.Repositories;
+
+namespace EnlightenmentApp.DAL.Repositories
+{
+    public class ChapterRepository : GenericRepository<ChapterEntity>, IChapterRepository
+    {
+        public ChapterRepository(DatabaseContext dbContext) : base(dbContext) 
+        {
+            this._context = dbContext;
```
Comment (Finka95) [inline Client/EnlightenmentApp.DAL/Repositories/ChapterRepository.cs:11]: You can remove "this._context = dbContext;"

File: Client/EnlightenmentApp.DAL/Repositories/GenericRepository.cs [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+                await _context.SaveChangesAsync(ct);
+                return entity;
+            }
+
+            throw new ArgumentNullException();
+        }
+
+        public virtual async Task<TEntity> Update(TEntity entity, CancellationToken ct)
+        {
+            if (await EntityExists(entity, ct))
```
Comment (Finka95) [inline Client/EnlightenmentApp.DAL/Repositories/GenericRepository.cs:50]: You should move this check into Business layer

File: Client/EnlightenmentApp.DAL/Repositories/GenericRepository.cs [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+                await _context.SaveChangesAsync(ct);
+                return entity;
+            }
+
+            throw new DbUpdateConcurrencyException();
+        }
+
+        public virtual async Task<bool> Delete(int id, CancellationToken ct)
+        {
+            var result = await _dbSet.FindAsync(id, ct);
```
Comment (Finka95) [inline Client/EnlightenmentApp.DAL/Repositories/GenericRepository.cs:63]: To Business layer too

File: Client/EnlightenmentApp.DAL/Repositories/ModuleRepository.cs [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+            {
+                return module;
+            }
+
+            throw new KeyNotFoundException();
+        }
+
+        public override async Task<ModuleEntity> Add(ModuleEntity moduleEntity, CancellationToken ct)
+        {
+            if (moduleEntity != null)
```
Comment (Finka95) [inline Client/EnlightenmentApp.DAL/Repositories/ModuleRepository.cs:31]: to Business layer too

File: Client/EnlightenmentApp.DAL/Repositories/ModuleReviewRepository.cs [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+using EnlightenmentApp.DAL.Interfaces.Repositories;
+using Microsoft.EntityFrameworkCore;
+
+namespace EnlightenmentApp.DAL.Repositories
+{
+    public class ModuleReviewRepository : GenericRepository<ModuleReviewEntity>, IModuleReviewRepository
+    {
+        public ModuleReviewRepository(DatabaseContext dbContext) : base(dbContext)
+        {
+            this._context = dbContext;
```
Comment (Finka95) [inline Client/EnlightenmentApp.DAL/Repositories/ModuleReviewRepository.cs:12]: You can remove "this..."

**Review by Finka95 — CHANGES_REQUESTED** [review]

File: Client/EnlightenmentApp.DAL/Repositories/GenericRepository.cs (hunk 6 above)
Comment (VadimZharikov) [reply Client/EnlightenmentApp.DAL/Repositories/GenericRepository.cs:50]: will be externalized in bll branch

File: Client/EnlightenmentApp.DAL/Repositories/GenericRepository.cs (hunk 7 above)
Comment (VadimZharikov) [reply Client/EnlightenmentApp.DAL/Repositories/GenericRepository.cs:63]: will be externalized in bll branch

File: Client/EnlightenmentApp.DAL/Repositories/ModuleRepository.cs (hunk 8 above)
Comment (VadimZharikov) [reply Client/EnlightenmentApp.DAL/Repositories/ModuleRepository.cs:31]: same

**Review by Finka95 — APPROVED** [review]

**Review by Glipchik — APPROVED** [review]

**Pull request merged**